File: crs/src/crs-cp-linux/fuzzer/SkyTracer/skytracer.go
package main

import (
	"bytes"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

func printSeperator() {
	for i := 0; i < 40; i++ {
		fmt.Print("*")
	}
	fmt.Println()
}

func runCommand(cmd *exec.Cmd, desc string) {
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	printSeperator()
	fmt.Printf("Command: %s\n", cmd.String())
	err := cmd.Run()

	if err != nil {
		fmt.Printf("Error running command: %s \n", err)
		fmt.Print(stderr.String())
		fmt.Printf("skytracer failed at stage: %s \n", desc)
		fmt.Println("Exiting...")
		printSeperator()
		os.Exit(-1)
	}

	fmt.Println("Command Output:")
	fmt.Println(stdout.String())
	printSeperator()
}

var TRACER_INSTRUMENT string = `
static void skytracer_init(int argc, char **argv, char **envp) {
asm("mov $0xf000, %rax; syscall");
}

static void skytracer_fini(void) {
asm("mov $0xf001, %rax; syscall");
}

__attribute__((section(".init_array"), used)) static typeof(skytracer_init) *init_p = skytracer_init;
__attribute__((section(".fini_array"), used)) static typeof(skytracer_fini) *fini_p = skytracer_fini;
`

func createPatchedTestHarness(testHarness, patchedTestHarness string) string {
	source, err := os.ReadFile(testHarness)
	if err != nil {
		fmt.Println("Error reading test harness src:", err)
		os.Exit(-1)
	}

	newSource := string(source) + TRACER_INSTRUMENT

	err = os.WriteFile(patchedTestHarness, []byte(newSource), 0644)
	if err != nil {
		fmt.Println("Error creating patched test harness:", err)
		os.Exit(-1)
	}

	cmd := exec.Command("gcc", patchedTestHarness, "-static", "-o", patchedTestHarness[:len(patchedTestHarness)-2])
	runCommand(cmd, "Build test harness")

	fmt.Printf("Patched test harness src created at : %s \n", patchedTestHarness)

	return patchedTestHarness[:len(patchedTestHarness)-2]
}

func createSymtabs(vmlinux string) string {
	vmlinuxSymtabs := fmt.Sprintf("%s.symtabs", vmlinux)

	vmlinuxFileInfo, err := os.Stat(vmlinux)
	if err != nil {
		fmt.Printf("vmlinux file does not exist: %s\n", err)
		os.Exit(-1)
	}

	symtabsFileInfo, err := os.Stat(vmlinuxSymtabs)
	if err == nil {
		vmlinuxTime := vmlinuxFileInfo.ModTime()
		symtabsTime := symtabsFileInfo.ModTime()
		if vmlinuxTime.Before(symtabsTime) {
			fmt.Println("No need to create new symtabs file")
			return vmlinuxSymtabs
		}
	}

	f, err := os.Create(vmlinuxSymtabs)
	if err != nil {
		fmt.Println("Error creating symtabs file:", err)
		os.Exit(-1)
	}
	defer f.Close()

	cmd := exec.Command("objdump", "-t", vmlinux)
	output, err := cmd.Output()
	if err != nil {
		fmt.Println("Error running objdump:", err)
		os.Exit(-1)
	}

	lines := strings.Split(string(output), "\n")
	lines = lines[1 : len(lines)-1]
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if !strings.Contains(line, ".text") {
			continue
		}
		tokens := strings.Fields(line)
		if len(tokens) < 2 {
			continue
		}
		addr := tokens[0]
		name := tokens[len(tokens)-1]
		fmt.Fprintf(f, "%s, %s\n", name, addr)
	}
	return vmlinuxSymtabs
}

func runVirtme(linuxKernel, qemuBin, workdir, vmlinuxSymtabs, traceOutputPath, patchedHarness, testBlob string) {
	testBlobFilename := filepath.Base(testBlob)

	cmd := exec.Command("cp", testBlob, workdir+testBlobFilename)
	runCommand(cmd, "Copy blob to workdir")

	script := workdir + "trace.sh"
	file, err := os.Create(script)
	if err != nil {
		fmt.Println("Error creating script file:", err)
		os.Exit(-1)
	}

	fmt.Fprintln(file, "#!/bin/bash")
	fmt.Fprintln(file, patchedHarness+" "+workdir+testBlobFilename)
	file.Close()

	cmd = exec.Command("chmod", "755", script)
	runCommand(cmd, "Chmod script file to execute")

	cmdAndArgs := []string{"virtme-run", "--memory", "2G", "--mods=auto", "--kopt", "panic=-1"}
	cmdAndArgs = append(cmdAndArgs, "--kopt", "nokaslr")
	cmdAndArgs = append(cmdAndArgs, "--kimg", linuxKernel)
	cmdAndArgs = append(cmdAndArgs, "--qemu-bin", qemuBin)
	cmdAndArgs = append(cmdAndArgs, "--rwdir", workdir+"="+workdir)
	cmdAndArgs = append(cmdAndArgs, "--disable-kvm")
	cmdAndArgs = append(cmdAndArgs, "--script-sh", script)
	cmd = exec.Command(cmdAndArgs[0], cmdAndArgs[1:]...)
	cmd.Env = append(os.Environ(), "SYMTABS="+vmlinuxSymtabs, "SKYTRACE_OUT="+traceOutputPath)
	runCommand(cmd, "virtme-run with script")
}

func main() {
	var (
		testHarnessName = flag.String("harness", "", "test harness src file name")
		testBlobName    = flag.String("blob", "", "test blob name")
	)

	flag.Parse()
	flag.VisitAll(func(f *flag.Flag) {
		if f.DefValue == f.Value.String() {
			fmt.Printf("Error: Flag %s is required.\n", f.Name)
			os.Exit(-1)
		}
	})

	testHarnessPrefix := "/cp-linux/src/test_harnesses/"
	testHarness := testHarnessPrefix + *testHarnessName

	testBlobPrefix := "/cp-linux/exemplar_only/blobs/"
	testBlob := testBlobPrefix + *testBlobName

	traceDir := "/tracer/"
	trace := traceDir + *testHarnessName + ".trace"
	patchedHarnessSrc := traceDir + "trace_" + *testHarnessName

	kernelPrefix := "/src-no-kasan/"
	vmlinux := kernelPrefix + "vmlinux"

	err := os.MkdirAll(traceDir, os.ModePerm)
	if err != nil {
		fmt.Println("Error creating trace work directory:", err)
		os.Exit(-1)
	}

	patchedHarness := createPatchedTestHarness(testHarness, patchedHarnessSrc)
	vmlinuxSymtabs := createSymtabs(vmlinux)
	linuxKernel := kernelPrefix + "arch/x86/boot/bzImage"
	qemuBin := "/fuzzer/SkyTracer/out/qemu-system-x86_64"
	runVirtme(linuxKernel, qemuBin, traceDir, vmlinuxSymtabs, trace, patchedHarness, testBlob)
}

File: crs/src/crs-cp-linux/fuzzer/SkyTracer/skytracer.py
#!/usr/bin/env python3
import argparse
import datetime
import os
import subprocess
import time
import shutil
from pathlib import Path
import pwn
import threading

if __name__ == "__main__":
    from trace import load_trace
else:
    from .trace import load_trace


SKYTRACER = Path(os.path.dirname(__file__)) / "out" / "qemu-system-x86_64"
PRE_SHELL = b"root@(none):/#"
GUEST_DIR = Path("/tmp/tracer/")
TRACER_INSTRUMENT = """
static void skytracer_init(int argc, char **argv, char **envp) {
  asm("mov $0xf000, %rax; syscall");
}

static void skytracer_fini(void) {
  asm("mov $0xf001, %rax; syscall");
}

__attribute__((section(".init_array"), used)) static typeof(skytracer_init) *init_p = skytracer_init;
__attribute__((section(".fini_array"), used)) static typeof(skytracer_fini) *fini_p = skytracer_fini;
"""
TIMEOUT=60
HANG_TIMEOUT=5
MONITOR_INTERVAL=1

def load_symtabs(path):
    ret = {}
    with open (path, "rt") as f:
        for line in f.read().split("\n")[:-1]:
            tmp = line.split(",")
            name = tmp[0]
            addr = int(tmp[1], 16)
            ret[addr] = name
    return ret

class SkyTracer:
    def __init__(self, kernel, workdir, monitor=False):
        os.makedirs(workdir, exist_ok = True)
        self.workdir = Path(workdir)
        self.out_path = self.workdir / "trace"
        self.kernel = Path(kernel)
        self.symtabs = self.kernel / "vmlinux.symtabs"
        if not self.symtabs.exists():
            raise Exception(f"{self.symtabs} does not exists")
        self.running = False
        if monitor:
            self.monitor = threading.Thread(target=self.monitor)
            self.monitor.start()

    def __boot_kernel(self):
        args  = ["virtme-run", "--memory 2G", "--mods=auto", "--kopt", "panic=-1"]
        args += ["--kopt", "nokaslr"]
        args += ["--kimg", str(self.kernel / "arch/x86/boot/bzImage")]
        args += ["--qemu-bin", str(SKYTRACER)]
        args += ["--rwdir", f"{GUEST_DIR}={self.workdir}"]
        args += ["--disable-kvm"]
        args += ["--no-virtme-ng-init"]

        cmd = " ".join(args)
        env = dict(os.environ)
        env["SYMTABS"] = str(self.symtabs)
        env["SKYTRACE_OUT"] = str(self.out_path)
        proc = pwn.process(cmd, shell=True, env=env)
        proc.recvuntil(PRE_SHELL)
        return proc

    def __run(self, cmd, timeout=TIMEOUT):
        if self.proc.poll() is not None:
            self.proc = self.__boot_kernel()
        if isinstance(cmd, str): cmd = bytes(cmd, "utf-8")
        cmd = f"timeout {timeout} ".encode() + cmd
        self.running = True
        self.proc.sendline(cmd)
        output = self.proc.recvuntil(PRE_SHELL)[:-len(PRE_SHELL)]
        self.running = False
        return output

    def prepare_target(self, target_src):
        target_src = Path(target_src)
        name = target_src.name
        if not name.endswith(".c"):
            raise Exception(f"target ({target_src}) must endwith `.c`")
        with open(target_src, "rt") as f : code = f.read()
        code += TRACER_INSTRUMENT
        target_src = self.workdir / name
        with open(target_src, "wt") as f: f.write(code)
        target_bin = self.workdir / name[:-2]
        print("[WARNING] TODO: We must use build.sh later")
        os.system(f"gcc {target_src} -static -o {target_bin}")
        if not target_bin.exists():
            raise Exception(f"Compile {target_src} fail")

        return target_bin

    def trace_target(self, target_bin, blob: bytes, timeout=TIMEOUT):
        self.run_target(target_bin, blob, timeout=timeout)
        trace = load_trace(self.out_path)
        return trace

    def run_target(self, target_bin, blob: bytes, timeout=TIMEOUT):
        target_bin = Path(target_bin)
        if not target_bin.exists():
            raise Exception(f"Execute prepare_target first!: {target_bin} does not exist")
        try:
            shutil.copy2(target_bin, self.workdir / target_bin.name)
        except shutil.SameFileError:
            pass
        blob_path = self.workdir / "blob"
        with open (blob_path, "wb") as f: f.write(blob)
        try:
            os.remove(self.out_path)
        except FileNotFoundError:
            pass
        cmd = f"{GUEST_DIR / target_bin.name} {GUEST_DIR / blob_path.name}"
        try:
            print(self.__run(cmd, timeout=timeout))
        except Exception as e:
            print(e)
        os.remove(blob_path)

    def __enter__(self):
        self.proc = self.__boot_kernel()
        return self

    def __exit__(self, exc_type, exc_value, exc_traceback):
        proc = self.proc
        sid = os.getsid(proc.proc.pid)
        while proc.poll() is None:
            os.system(f"pkill -s {sid}")
            time.sleep(0.1)

    def monitor(self):
        while True:
            try:
                time.sleep(MONITOR_INTERVAL)
                if not self.running:
                    continue
                hang_time = time.time() - self.out_path.stat().st_mtime
                print(f"[MONITOR] Hang for {hang_time}s ...")
                if hang_time < HANG_TIMEOUT:
                    continue
                print("[MONITOR] Killing ...")
                self.running = False
                self.__exit__(None, None, None)
            except Exception as e:
                print(e)
                pass

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("kernel",
                        help="path to the kernel directory for SkyTracer")
    parser.add_argument("target",
                        help="path to the target source or binary")
    parser.add_argument("--workdir", default="/tmp/tracer",
                        help="path to the working directory")
    parser.add_argument("--no_load", action="store_true")
    parser.add_argument("--timeout", default=TIMEOUT)
    parser.add_argument("--monitor", action="store_true")
    return parser.parse_args()

if __name__ == "__main__":
    args = parse_args()
    print(args)
    print("Starting SkyTracer ...")
    with SkyTracer(args.kernel, args.workdir, args.monitor) as t:
        funcs = load_symtabs(t.symtabs)
        target = args.target
        if target.endswith(".c"):
            target = t.prepare_target(target)
        print("SkyTracer is ready")
        while True:
            blob = input("Enter the blob path:\n")
            with open (blob, "rb") as f: data = f.read()
            t.run_target(target, data, timeout=args.timeout)
            print("Here is the trace path:")
            print(t.out_path)
            if not args.no_load:
                trace = load_trace(t.out_path)
                os.remove(t.out_path)
                print(datetime.datetime.now())
                print(trace.str(funcs))
